### DISCUSSION TITLE
Clambda needed to be in lambda list of MBAR in Amber

### DISCUSSION THREAD

**Discussion by adamkni** [alchemistry/alchemlyb#298] [Q&A]
I am trying to do calculations of both MBAR and TI in one run, however, the Amber parser does not allow to have clambda different from the set of lambda values for MBAR. (amber.py line 239-246) Could you explain to me please why is that needed? I thought clambda is for TI calculations and MBAR lambdas only for MBAR calculations and they do not need to have same values. 
Thanks.

**Comment by orbeckst** [comment]
@DrDomenicoMarson do you know the answer to this AMBER question?

**Reply by DrDomenicoMarson** [reply]
I'm not aware of a theoretical background for this requirement, it seemed strange to me when I first looked at the code! I think this is a "feature" we inherited when ported the code from the old library, as it was coded in this way.
I could try to get rid of this limitation, but I don't know when I could work on this unfortunately!

PS: when I first encountered this issue I just ran the calculations with more lambdas, I used the same number of lambda windows for TI and MBAR, usually the more the better for both, so it's not a real limitation with current computing power, I think...

**Comment by xiki-tempula** [comment]
Yes, I do agree that in theory one doesn't need the current lambda to be in the set of mbar values to be computed. We kind of made this lazy assumption that one set the lambdas to be computed and then run a separate window for each lambda windows.

I'm a bit curious as to why one would need to not do this? I'm meaning not computing the reduced potential for the current window as it is free.

**Reply by adamkni** [reply]
Well, I was running TI calculations, with already lambda windows set and then I decided to add also the possibility to analyse it for MBAR later. And since I have seen mainly uniformly spaced windows in MBAR, I just put those, but it was not possible to analyze them, because they were not the same as TI windows. 

It would be nice to mention this assumption in the docs if it is not going to be corrected soon so that no one gets surprised later.

**Reply by xiki-tempula** [reply]
I see, in this case, I would recommend setting the MBAR windows to be the same as the TI windows. 
For MBAR calculation to give good estimate, one need to make sure that, at least, the adjacent MBAR windows would have good overlap. A plotting function was provided to check this https://alchemlyb.readthedocs.io/en/latest/visualisation.html#plot-overlap-matrix. If the MBAR windows are not the same as the TI windows, it would makes this quality assessment difficult.

**Comment by adamkni** [comment]
@xiki-tempula @DrDomenicoMarson Also I have seen in tests, that not all of the lambdas in BAR calculations were included in TI calculations. So I added a few more lambdas to the BAR lambdas, however, I found that the code does not allow that for BAR calculations, however, it is completely fine for MBAR calculations. Is there any specific reason for that or is that again an artifact from the old code? 

The problem in the code is in the estimators bar_.py at lines 106 and 111, where it asks for the groups, however, it does not check if some value in the N_k array is zero. And all despite the fact, that the code itself gives to N_k value of zero which inevitably leads to errors.

**Reply by xiki-tempula** [reply]
Sorry, I struggled to understand your concern.

Is your problem that you have done a couple of calculations, where at each lambda state, you run the TI calculations and you have computed the MBAR values for all lambda values. You have trouble calculating the BAR estimate but the MBAR is fine?

**Reply by adamkni** [reply]
Yes and the problem seems in the fact, that for BAR calculations, the TI lambda values need to be **exactly** the same as the BAR lambdas. However, for MBAR, it does not seem necessary. I am doing that because, for TI, it is not recommended to have lambda values at 0 and 1, however for BAR, it is recommended.

**Reply by xiki-tempula** [reply]
I don't think you could do BAR if your TI window and MBAR window doesn't match? For MBAR, in theory, one could run the estimate as it uses the reduced potential between all lambda windows. In practice, it is more problematic as one cannot do decorrelation as it is based on `dE`.
But I don't think you can run BAR as BAR only takes into account the forward and backward dE. So you need the reduced potential for the next and precious lambda window. If your TI window doesn't match the MBAR window, the forward and backward dE is ill-defined?
